crates/libs/result/src/error.rs

```diff
@@ -1,63 +1,265 @@
+/// `IErrorInfo` object. The `IErrorInfo` object is a COM object that can provide detailed information
+/// about an error, such as a text string, a `ProgID` of the originator, etc. If the error object
+/// was originated in an WinRT component, then additional information such as a stack track may be
+/// captured.
+///
+/// However, many systems based on COM do not use `IErrorInfo`. For these systems, the optional error
+/// info within `Error` has no benefits, but has substantial costs because it increases the size of
+/// the `Error` object, which also increases the size of `Result<T>`.
+///
+/// The `slim-errors` Cargo feature allows you to _statically_ disable error detail in `Error`.
+/// This allows the `Error` and even the `Result<(), Error>` types to have a very small size; they
+/// have the same size and alignment as `HRESULT` itself. Because of this, they can be returned
+/// directly in a machine register, which gives better performance than writing the (relatively
+/// large) `Error` object (with full error detail) to the stack on every function return.
+#[derive(Clone)]
+pub struct ErrorT<Detail = DefaultDetail> {
+    /// The error code. Because this uses `NonZeroHRESULT`, this provides a "niche" to the Rust
+    /// compiler. This allows the compiler to use more compact representations in some sitatutions.
+    code: NonZeroHRESULT,
+
+    /// Contains details about the error, such as error text.
+    ///
+    /// Note; We cannot use `Option<Detail>` here, because that would increase the size of `Error`
+    /// even when `Detail` contains no information at all (a ZST).  The `Detail` type itself must
+    /// be a ZST for `Error` to be a slim type.
+    detail: Detail,
+}
 
 /// An error object consists of both an error code as well as detailed error information for debugging.
-#[derive(Clone, PartialEq, Eq)]
-pub struct Error {
-    code: HRESULT,
-    #[cfg(windows)]
-    info: Option<ComPtr>,
+///
+/// This is a type alias for the [`ErrorT`] type.
+pub type Error = ErrorT<DefaultDetail>;
+
+/// Specifies the default error detail for the `Error` type.
+#[cfg(not(feature = "slim-errors"))]
+pub type DefaultDetail = ErrorInfo;
+
+/// Specifies the default error detail for the `Error` type.
+#[cfg(feature = "slim-errors")]
+pub type DefaultDetail = NoDetail;
+
+/// On non-Windows platforms, `ErrorInfo` is mapped to `NoDetail` because non-Windows platforms
+/// do not support `IErrorInfo`.
+#[cfg(not(windows))]
+pub type ErrorInfo = NoDetail;
+
+/// This type implements `ErrorDetail`, by containing no error information at all. This type allows
+/// for "slim" `Error` objects, which have the same size as `HRESULT`.
+#[derive(Default, Clone)]
+pub struct NoDetail;
+
+/// Defines the behavior of error detail objects, which are stored within `Error`.
+pub trait ErrorDetail: Sized {
+    /// Returns an empty error detail.
+    fn empty() -> Self;
```

Comment:
Not much different, but this way everything is defined on a single trait.  Otherwise I'll need to add `+ Default` to lots of places in this file, for not much real benefit.

Reply:
Could you not use `pub trait ErrorDetail: Sized + Default {` ?

Reply:
Yeah, that would work.  I'll switch to that.

Reply:
Done